js/viz.js

```diff
@@ -1,29 +1,101 @@
 // eslint-disable-next-line no-unused-vars
 const viz = {
-  draw(context, websites) {
-    const pi = Math.PI * 2;
-    let x = 50, y = 50;
-
-    context.fillStyle = 'white';
-    context.beginPath();
-
-    for (const website in websites) {
-      context.moveTo(x, y);
-      context.arc(x, y, 10, 0, pi);
-      context.fillText(website, x, y);
-
-      let x1 = x;
-      for (const thirdParty in websites[website].thirdPartyRequests) {
-        x1+=150;
-        context.moveTo(x1, y);
-        context.arc(x1, y, 5, 0, pi);
-        context.fillText(thirdParty, x1, y);
-      }
-
-      x+=20; y+=20;
-    }
-
-    context.stroke();
-    context.fill();
+  init(nodes, links) {
+    const svg = d3.select('svg');
+    const nodesGroup = svg.append('g');
+    nodesGroup.attr('class', 'nodes');
+    const linksGroup = svg.append('g');
+    linksGroup.attr('class', 'links');
+    this.allCircles = nodesGroup.selectAll('.node');
+    this.allLines = linksGroup.selectAll('.link');
+
+    // D3 needs explicit pixel values for 'center_force'
+    const visualization = document.getElementById('visualization');
+    const width = visualization.getBoundingClientRect().width;
+    const height = visualization.getBoundingClientRect().height;
+
+    this.simulation = d3.forceSimulation(nodes);
+    this.simulation.force('charge', d3.forceManyBody());
+    const linkForce = d3.forceLink(links);
+    // link source/target values are uniquely identified by
+    // the hostname property
+    linkForce.id(function (d) {
+      return d.hostname;
+    });
+    linkForce.distance(50);
+    this.simulation.force('link', linkForce);
+    this.simulation.force('center', d3.forceCenter(width/2, height/2));
```

Comment:
Is there any way to prevent this pinging off the graph. Mine zoomed out of sight pretty quickly... can be a follow up.